src/test/java/ua/bookstore/online/service/impl/BookServiceImplTest.java

```diff
@@ -0,0 +1,401 @@
+import static ua.bookstore.online.utils.ConstantAndMethod.AUTHOR;
+import static ua.bookstore.online.utils.ConstantAndMethod.CATEGORY_IDS;
+import static ua.bookstore.online.utils.ConstantAndMethod.ID_1;
+import static ua.bookstore.online.utils.ConstantAndMethod.ISBN;
+import static ua.bookstore.online.utils.ConstantAndMethod.PRICE;
+import static ua.bookstore.online.utils.ConstantAndMethod.TITLE;
+import static ua.bookstore.online.utils.ConstantAndMethod.createBook;
+import static ua.bookstore.online.utils.ConstantAndMethod.createBookRequestDto;
+import static ua.bookstore.online.utils.ConstantAndMethod.getBookDto;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+import ua.bookstore.online.dto.book.BookDto;
+import ua.bookstore.online.dto.book.BookDtoWithoutCategoryIds;
+import ua.bookstore.online.dto.book.CreateBookRequestDto;
+import ua.bookstore.online.exception.EntityNotFoundException;
+import ua.bookstore.online.exception.UniqueIsbnException;
+import ua.bookstore.online.mapper.BookMapper;
+import ua.bookstore.online.model.Book;
+import ua.bookstore.online.repository.book.BookRepository;
+import ua.bookstore.online.repository.book.BookSpecificationBuilder;
+import ua.bookstore.online.service.CategoryService;
+
+@ExtendWith(MockitoExtension.class)
+class BookServiceImplTest {
+    @InjectMocks
+    private BookServiceImpl bookService;
+    @Mock
+    private BookMapper bookMapper;
+    @Mock
+    private BookRepository bookRepository;
+    @Mock
+    private CategoryService categoryService;
+    @Mock
+    private BookSpecificationBuilder bookSpecificationBuilder;
+
+    @Test
+    @DisplayName("Save new Book and return BookDto")
+    void save_WithUniqueIsbnAndValidCategories_ReturnsBookDto() {
+        // Given
+        CreateBookRequestDto requestDto = createBookRequestDto();
+        Book book = createBook();
+
+        // Mocking behavior
+        mockingForSaveMethod(requestDto, CATEGORY_IDS);
+        when(bookRepository.save(any(Book.class))).thenReturn(book);
+        mockingMapperMethods(book);
+
+        // When
+        BookDto actual = bookService.save(requestDto);
+
+        // Then
+        assertNotNull(actual);
+        BookDto expected = getBookDto(book);
+        assertTrue(EqualsBuilder.reflectionEquals(expected, actual));
+
+        // Verify repository method calls
+        verify(bookRepository).findByIsbn(requestDto.isbn());
```

Comment:
It's not necessary to verify execution of methods which you previously mocked by `when().then()`. `verify` is only useful when you don't mock the method to return desired result but simply want to verify that method was actually called. For more details you can check the javadoc of `verify` method.